EventListener/IsGrantedListener.php

```diff
@@ -0,0 +1,100 @@
+use Symfony\Component\Security\Core\Exception\AccessDeniedException;
+
+/**
+ * Handles the IsGranted annotation on controllers.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class IsGrantedListener implements EventSubscriberInterface
+{
+    private $authChecker;
+
+    public function __construct(AuthorizationCheckerInterface $authChecker = null)
+    {
+        $this->authChecker = $authChecker;
+    }
+
+    public function onKernelController(FilterControllerEvent $event)
+    {
+        $request = $event->getRequest();
+        /** @var $configurations IsGranted[] */
+        if (!$configurations = $request->attributes->get('_is_granted')) {
+            return;
+        }
+
+        if (null === $this->authChecker) {
+            throw new \LogicException('To use the @IsGranted tag, you need to install symfony/security-bundle.');
+        }
+
+        foreach ($configurations as $configuration) {
+            $subject = null;
+            if ($configuration->getSubject()) {
+                if (!$request->attributes->has($configuration->getSubject())) {
+                    $allAttributes = $request->attributes->all();
+                    // remove one attribute we know is noise in the error message
+                    unset($allAttributes['_is_granted']);
+                    throw new \RuntimeException(sprintf('Could not find the subject "%s" for the @IsGranted annotation. Available attributes are: %s', $configuration->getSubject(), implode(', ', array_keys($allAttributes))));
+                }
+
+                $subject = $request->attributes->get($configuration->getSubject());
+            }
+
+            if (!$this->authChecker->isGranted($configuration->getAttributes(), $subject)) {
+                $argsString = $this->getIsGrantedString($configuration);
+
+                $message = $configuration->getMessage() ?: sprintf('Access Denied by controller annotation @IsGranted(%s)', $argsString);
+
+                if ($statusCode = $configuration->getStatusCode()) {
+                    throw new HttpException($statusCode, $message);
+                }
+
+                throw new AccessDeniedException($message);
+            }
+        }
+    }
+
+    private function getIsGrantedString(IsGranted $isGranted)
+    {
+        $attributes = array_map(function ($attribute) {
+            return sprintf('"%s"', $attribute);
+        }, (array) $isGranted->getAttributes());
+        if (1 === count($attributes)) {
+            $argsString = reset($attributes);
+        } else {
+            $argsString = sprintf('[%s]', implode(', ', $attributes));
+        }
+
+        if (null !== $isGranted->getSubject()) {
+            $argsString = sprintf('%s, %s', $argsString, $isGranted->getSubject());
+        }
+
+        return $argsString;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return array(KernelEvents::CONTROLLER => 'onKernelController');
```

Comment:
this is bad news. It should use the `CONTROLLER_ARGUMENT` event, to be compatible with argument resolvers (here, it would vote on the id of the Post, not on the Post object, when using the Symfony 3.x argument resolvers rather than the ParamConverter layer which has not been migrated to become an argument resolver precisely because of the Security annotation not being migrated to the new event yet)